frontend/projects/upgrade/src/app/core/stratification-factors/store/stratification-factors.effects.spec.ts

```diff
@@ -0,0 +1,171 @@
+import { StratificationFactor } from './stratification-factors.model';
+
+describe('StratificationFactorsEffects', () => {
+  let actions$: ActionsSubject;
+  let store$: any;
+  let stratificationFactorsDataService: any;
+  let router: any;
+  let service: StratificationFactorsEffects;
+  const mockData: StratificationFactor = {
+    factor: 'favourite_food',
+    factorValue: { pizza: 10, burger: 5 },
+    experimentIds: [],
+  };
+
+  beforeEach(() => {
+    actions$ = new ActionsSubject();
+    store$ = new BehaviorSubject({});
+    store$.dispatch = jest.fn();
+    stratificationFactorsDataService = {};
+    router = {
+      navigate: jest.fn(),
+    };
+
+    service = new StratificationFactorsEffects(actions$, router, store$, stratificationFactorsDataService);
+  });
+
+  describe('fetchStratificationFactors$', () => {
+    it('should dispatch actionFetchStratificationFactorsSuccess on API call success', fakeAsync(() => {
+      stratificationFactorsDataService.fetchStratificationFactors = jest.fn().mockReturnValue(of([mockData]));
+      selectAllStratificationFactors.setResult([mockData]);
+
+      const expectedAction = StratificationFactorsActions.actionFetchStratificationFactorsSuccess({
+        stratificationFactors: [mockData],
+      });
+
+      service.fetchStratificationFactors$.subscribe((result) => {
+        expect(result).toEqual(expectedAction);
+      });
+
+      actions$.next(StratificationFactorsActions.actionFetchStratificationFactors({ isLoading: true }));
+
+      tick(0);
+    }));
+
+    it('should dispatch actionFetchStratificationFactorsFailure on API call failure', fakeAsync(() => {
+      stratificationFactorsDataService.fetchStratificationFactors = jest
+        .fn()
+        .mockReturnValue(throwError(() => new Error('test')));
+      const expectedAction = StratificationFactorsActions.actionFetchStratificationFactorsFailure();
+
+      service.fetchStratificationFactors$.subscribe((result) => {
+        expect(result).toEqual(expectedAction);
+      });
+
+      actions$.next(StratificationFactorsActions.actionFetchStratificationFactors({ isLoading: true }));
+
+      tick(0);
+    }));
+  });
+
+  describe('deleteStratificationFactor$', () => {
+    it('should dispatch actionDeleteStratificationFactorSuccess and navigate to participants page on success', fakeAsync(() => {
+      stratificationFactorsDataService.deleteStratificationFactor = jest.fn().mockReturnValue(of([{ ...mockData }]));
+
+      const expectedAction = StratificationFactorsActions.actionDeleteStratificationFactorSuccess({
+        stratificationFactor: { ...mockData },
+      });
+
+      service.deleteStratificationFactor$.subscribe((result) => {
+        expect(result).toEqual(expectedAction);
+        expect(router.navigate).toHaveBeenCalledWith(['/participants']);
+      });
+
+      actions$.next(
+        StratificationFactorsActions.actionDeleteStratificationFactor({ factor: { ...mockData[0] }.factor })
+      );
```

Comment:
Firstly, thank you for writing unit tests, that's great. I found a weird thing here though, I found that this test is not actually running the expectations in the `subscribe` block. If I change `expectedAction` to null or `/participants` to something else, it still passes.

This is because the observable never emits. I'd think that this would fail the test, but apparently it does not, which is very confusing but here's what I found is happening:

The reason the observable never emits is that you're passing in the data param like this when dispatching `actionDeleteStratificationFactor`:

```
{ factor: { ...mockData[0] }.factor }
```

Which makes no sense, mockData is an object, so`.factor` is undefined here, and thus it is getting filtered out. you want `{ factor: mockData.factor}`.

When you pass the data in that way, the test will actually run the expect code in the `subscribe` block.

The reason I came into the unit tests in the first place is on the comment elsewhere about this, we should be able to use the unit tests here to test what happens when `data[0]` is undefined. When I test this, the navigation is not called and the action dispatched is `undefined`.

```
it('this should do something if the data coming back is unexpected or empty array, but it will neither navigate nor dispatch anything', fakeAsync(() => {
      stratificationFactorsDataService.deleteStratificationFactor = jest.fn().mockReturnValue(of([])); // <---return empty array from mock API response to simulate "success" but unexpected data.

      const expectedAction = StratificationFactorsActions.actionDeleteStratificationFactorSuccess({
        stratificationFactor: { ...mockData },
      });

      service.deleteStratificationFactor$.subscribe((result) => {
        expect(result).toEqual(expectedAction);
        expect(router.navigate).toHaveBeenCalledWith(['/participants']);
      });

      actions$.next(StratificationFactorsActions.actionDeleteStratificationFactor({ factor: mockData.factor }));

      tick(0);
    }));
```

Reply:
Okay, got it. Will add failure actions and update the backend to return the object instead of array of object.